### PR TITLE
[HttpFoundation] we should not pass size on FileBag removing the contruct parameter

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25466 
| License       | MIT
| Doc PR        | none

We may have forgotten this one, thanks to @craue for seeing it !

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/FileBag.php [hunk 1]
```diff
@@ -84,7 +84,7 @@ protected function convertFileInformation($file)
                 if (UPLOAD_ERR_NO_FILE == $file['error']) {
                     $file = null;
                 } else {
-                    $file = new UploadedFile($file['tmp_name'], $file['name'], $file['type'], $file['size'], $file['error']);
+                    $file = new UploadedFile($file['tmp_name'], $file['name'], $file['type'], null, $file['error']);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/FileBag.php]: we should deprecate passing the arg in fact, isn't it?

Comment (Simperfit) [reply src/Symfony/Component/HttpFoundation/FileBag.php]: This is what we are doing. But I can't remove it until 5.0, or can I?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/FileBag.php]: I missed that on the merged PR!
Yet we can, and we must in fact.
Right now, we are forced to pass a "null", which is not the signature we want on v5

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpFoundation/FileBag.php]: (but we must provide a BC layer of course)

Comment (Simperfit) [reply src/Symfony/Component/HttpFoundation/FileBag.php]: let's do it then ;)

Comment (Simperfit) [reply src/Symfony/Component/HttpFoundation/FileBag.php]: done

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 2]
```diff
@@ -47,24 +47,27 @@ class UploadedFile extends File
...
      */
-    public function __construct(string $path, string $originalName, string $mimeType = null, int $size = null, int $error = null, bool $test = false)
+    public function __construct(string $path, string $originalName, string $mimeType = null, int $error = null, $test = false)
     {
         $this->originalName = $this->getName($originalName);
         $this->mimeType = $mimeType ?: 'application/octet-stream';
-        $this->size = $size;
-        if (null !== $size) {
+
+        $this->error = isset(func_get_args()[4]) && is_int(func_get_args()[4]) ? func_get_args()[4] : $error;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: my patch proposal for this file:
```diff
--- a/src/Symfony/Component/HttpFoundation/File/UploadedFile.php
+++ b/src/Symfony/Component/HttpFoundation/File/UploadedFile.php
@@ -27,7 +27,6 @@ class UploadedFile extends File
     private $test = false;
     private $originalName;
     private $mimeType;
-    private $size;
     private $error;
 
     /**
@@ -47,7 +46,6 @@ class UploadedFile extends File
      * @param string      $path         The full temporary path to the file
      * @param string      $originalName The original file name of the uploaded file
      * @param string|null $mimeType     The type of the file as provided by PHP; null defaults to application/octet-stream
-     * @param int|null    $size         The file size provided by the uploader
      * @param int|null    $error        The error constant of the upload (one of PHP's UPLOAD_ERR_XXX constants); null defaults to UPLOAD_ERR_OK
      * @param bool        $test         Whether the test mode is active
      *                                  Local files are used in test mode hence the code should not enforce HTTP uploads
@@ -55,13 +53,15 @@ class UploadedFile extends File
      * @throws FileException         If file_uploads is disabled
      * @throws FileNotFoundException If the file does not exist
      */
-    public function __construct(string $path, string $originalName, string $mimeType = null, int $size = null, int $error = null, bool $test = false)
+    public function __construct(string $path, string $originalName, string $mimeType = null, int $error = null, $test = false)
     {
         $this->originalName = $this->getName($originalName);
         $this->mimeType = $mimeType ?: 'application/octet-stream';
-        $this->size = $size;
-        if (null !== $size) {
-            @trigger_error('Passing a size in the constructor is deprecated since 4.1 and will be removed in 5.0. Use getSize() instead.', E_USER_DEPRECATED);
+
+        if (4 < func_num_args() ? !is_bool($test) : @filesize($path) === $error) {
+            @trigger_error(sprintf('Passing a size as 4th argument to the constructor of "%s" is deprecated since 4.1 and will be unsupported in 5.0.', __CLASS__), E_USER_DEPRECATED);
+            $error = $test;
+            $test = 5 < func_num_args() ? func_get_arg(5) : false;
         }
         $this->error = $error ?: UPLOAD_ERR_OK;
         $this->test = $test;
@@ -152,7 +152,7 @@ class UploadedFile extends File
     {
         @trigger_error(sprintf('"%s" is deprecated since 4.1 and will be removed in 5.0. Use getSize() instead.', __METHOD__), E_USER_DEPRECATED);
 
-        return $this->size;
+        return $this->getSize();
     }
 
     /**
```

Comment (Simperfit) [reply src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: patch applied

**Comment by Simperfit** [comment]
@nicolas-grekas should we update the test before merging this one ?

File: src/Symfony/Component/Form/Tests/CompoundFormTest.php [hunk 3]
```diff
@@ -630,21 +630,21 @@ public function testSubmitPostOrPutRequest($method)
...
 
         $this->assertEquals('Bernhard', $form['name']->getData());
         $this->assertEquals($file, $form['image']->getData());
 
         unlink($path);
     }
 
-    /**
-     * @dataProvider requestMethodProvider
+    /** @dataProvider requestMethodProvider
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/CompoundFormTest.php]: Why this change?

File: src/Symfony/Component/Form/Tests/CompoundFormTest.php [hunk 4]
```diff
@@ -630,21 +630,21 @@ public function testSubmitPostOrPutRequest($method)
...
      */
     public function testSubmitPostOrPutRequestWithEmptyRootFormName($method)
     {
         $path = tempnam(sys_get_temp_dir(), 'sf2');
         touch($path);
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/CompoundFormTest.php]: we can now drop the `touch()` call everywhere if I don't miss anything

File: src/Symfony/Component/HttpFoundation/File/UploadedFile.php [hunk 5]
```diff
@@ -47,22 +46,24 @@ class UploadedFile extends File
...
-    public function __construct(string $path, string $originalName, string $mimeType = null, int $size = null, int $error = null, bool $test = false)
+    public function __construct(string $path, string $originalName, string $mimeType = null, int $error = null, $test = false)
     {
         $this->originalName = $this->getName($originalName);
         $this->mimeType = $mimeType ?: 'application/octet-stream';
-        $this->size = $size;
-        if (null !== $size) {
-            @trigger_error('Passing a size in the constructor is deprecated since 4.1 and will be removed in 5.0. Use getSize() instead.', E_USER_DEPRECATED);
+
+        if (4 < func_num_args() ? !is_bool($test) : @filesize($path) === $error) {
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/File/UploadedFile.php]: I think we should avoid the `filesize()` call when `$error` is `null`

**Comment by nicolas-grekas** [comment]
Status: needs work

File: src/Symfony/Component/Form/Tests/CompoundFormTest.php (hunk 3 above)
Comment (Simperfit) [reply src/Symfony/Component/Form/Tests/CompoundFormTest.php]: Ill revert

File: src/Symfony/Component/Form/Tests/CompoundFormTest.php (hunk 4 above)
Comment (Simperfit) [reply src/Symfony/Component/Form/Tests/CompoundFormTest.php]: I needed to add some in order to make some test work, I don't now if we can remove them ill check

**Comment by Simperfit** [comment]
PR rebased and comments done

**Comment by Simperfit** [comment]
Status: Needs Review

File: src/Symfony/Component/HttpKernel/Tests/ClientTest.php [hunk 6]
```diff
@@ -120,6 +125,7 @@ public function testUploadedFile()
 
             $this->assertEquals('original', $file->getClientOriginalName());
             $this->assertEquals('mime/original', $file->getClientMimeType());
+            $this->assertEquals($file->getSize(), 123);
             $this->assertTrue($file->isValid());
```
Comment (xabbuh) [inline src/Symfony/Component/HttpKernel/Tests/ClientTest.php]: this assertion seems to fail currently

**Comment by nicolas-grekas** [comment]
Status: needs work

File: src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php [hunk 7]
```diff
@@ -60,6 +64,26 @@ public function testShouldSetEmptyUploadedFilesToNull()
         $this->assertNull($bag->get('file'));
     }
 
+    /**
+     * @group legacy
+     * @expectedDeprecation Passing a size as 4th argument to the constructor of "Symfony\Component\HttpFoundation\File\UploadedFile" is deprecated since Symfony 4.1 and will be unsupported in 5.0.
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php]: I think we should not make these tests expect a deprecation, but update it to not trigger it (i.e. make the `UploadedFile` constructor treat the passed integer as the error code instead of the file size).

File: src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php [hunk 8]
```diff
@@ -60,6 +64,26 @@ public function testShouldSetEmptyUploadedFilesToNull()
...
+     */
+    public function testShouldNotTriggerDeprecationWhenPassingSize()
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php]: I think we should remove this test entirely. Not even does the test name not match the asserted output, it is also just the repetition of what we already have in the `UploadedFileTest` class.

File: src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php (hunk 7 above)
Comment (xabbuh) [reply src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php]: see 7bb538bdca0b8c80dc781e6e2d4908179bd893d1

File: src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php (hunk 8 above)
Comment (xabbuh) [reply src/Symfony/Component/HttpFoundation/Tests/FileBagTest.php]: see 7bb538bdca0b8c80dc781e6e2d4908179bd893d1

**Comment by xabbuh** [comment]
@Simperfit FYI, I have pushed some changes that IMO fix some issues in the tests and make the build green

**Review by xabbuh — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]
(@xabbuh please squash your commits before merging)

**Comment by xabbuh** [comment]
@nicolas-grekas done

**Comment by nicolas-grekas** [comment]
Thank you @Simperfit.

**Comment by alexander-schranz** [comment]
We updated our project to symfony 4.1 and it did throw an error looks like a undocumented bc break in the constructor of UploadedFile?

/cc @nicolas-grekas @xabbuh

**Comment by nicolas-grekas** [comment]
There is a BC layer in place so that this shouldn't break anything.
Please open a dedicated issue with precise details about the error and a reproducer ideally.

**Comment by alexander-schranz** [comment]
@nicolas-grekas ok seems like the func_num_args check is false. #28527
